### DISCUSSION TITLE
Wasm alternative for WebView2

### DISCUSSION THREAD

**Discussion by gevart** [unoplatform/uno#13816] [Q&A] [answered]
After some time I need to use WebView2 control on Wasm once again. Last time I've created Wasm-specific control using HtmlElement attribute and iframe and it worked fine, but that was a couple of years ago so it was a shared project solution architecture and I could just create the control inside Wasm head while still using it inside Shared code with conditional XAML markup.

But now I'm using the latest Uno templates so I have multi-target library solution and it seems I can't do that. And to make things worse, I can't use conditional Wasm XAML markup too on .Net 7, as it's mentioned in the docs. So I wonder what is the modern way to create a control to display html on Wasm?

**Comment by jeromelaban** [comment]
Thanks for the question. The current state of this topic was supposed to be simplified with the introduction of `net8.0-browser`, but this was moved to another release of .NET.

There are two ways to handle this:
- Create a [cross runtime library](https://platform.uno/docs/articles/get-started-dotnet-new.html?tabs=net7%2Cwindows#uno-platform-blank-cross-runtime-library), which will allow you to build such a control in a platform independent way
- Move your control to a net7.0-only library (which can reference the required wasm-only nuget packages) that you reference only from the webassembly head, then use dependency injection (or simply an interface) that will allow your shared class library to create a control instance that can be injected in your visual tree using code.

**Reply by gevart** [reply]
@jeromelaban thanks for the quick reply!

I think I understand the 2nd way, will try it and report back.

As to the first one, just to clarify. If I create a separate cross runtime library, will I be able to use platform conditional XAML/code compilation for wasm/non-wasm targets to reuse WebView2 where possible? Or will I have to build universal control from scratch?

**Reply by jeromelaban** [reply]
Thanks for the update.

For the first one, you won't need conditional XAML in this case, as you the control is available for all platforms, but its implementation differs based on the platform (it's the bait-and-switch pattern).

**Comment by CalebSerafin** [comment, marked as answer]
I prefer the DI approach, although its may not be everyone's favourite.

I've created a fresh project with the Class Library structure.
Then in the class library I define a IWebView with an `void Open(Url)` method.
In the heads: I create implementations for that interface.
I then register those services as transient in the App's ServiceCollection.
Consumers can then get and use instances of the webView.

**Note: This works perfectly on Wasm, However, I could not get the native WinUI3 WebView2 to work, even as stand-alone (Direct reference in XAML, no DI shenanigans). This might just be an issue with my setup and for others, it should work.**

Repo and Branch: https://github.com/CalebSerafin/UnoSamples2023/tree/WebViewOnWasm

Commit to see changes: https://github.com/CalebSerafin/UnoSamples2023/compare/master...WebViewOnWasm

**Reply by gevart** [reply]
@CalebSerafin thanks for the input! I'm a DI enthusiast myself so I'd love to try your solution first but I probably don't have rights to see the repo
![image](https://github.com/unoplatform/uno/assets/11190768/f3ffe5ef-5394-4123-8c2e-a47394f91cf5)

**Reply by CalebSerafin** [reply]
@gevart Ah sorry about that. missed that Visual Studio defaults to private. Repo is now public 👍

**Reply by gevart** [reply]
@CalebSerafin thanks a lot, I tried your solution and it worked just fine!

I also managed to make it work on Windows with `WebView2`, the culprit in your sample was a separate `Program.cs` in the Windows head that you used to register Win `IWebView` implementation. If you remove it and register service in `AppHead.xaml.cs` instead (somewhere, just make sure it's called before building host) :

```
#if WINDOWS
        App.ConfigureServices((context, services) => {
            services.AddTransient<IWebView, WindowsWebView>();
        });
#endif
```

your sample should work on Windows too.

**Reply by CalebSerafin** [reply]
Ah great! Yeah I guess I didn't need to seperate the WebView xaml from the main source; since it should still compile when targeting other UNO Platforms.
